draft-ietf-tls-exported-authenticator.md

```diff
@@ -142,70 +175,103 @@ CertificateVerify
+certificate selection, with the extensions provided in the authenticator request
+taking precedence over the extensions provided in the connection handshake.
+
+If an authenticator request was provided, the Certificate message MUST contain
+only extensions present in the authenticator request. Otherwise, the Certificate
+message MUST contain only extensions present in the ClientHello.
+
+The authenticator message is the concatenation of messages:
 Certificate || CertificateVerify || Finished
 
-A given exported authenticator can be validated by checking the validity of the
-CertificateVerify message and recomputing the Finished message to see if it
-matches.
+A given authenticator can be validated by checking the validity of the
+CertificateVerify message given the authenticator request (if used) and recomputing the
+Finished message to see if it matches.
 
 # API considerations
 
-The creation and validation of exported authenticators SHOULD be implemented inside
+The creation and validation of both authenticator requests and
+authenticators SHOULD be implemented inside the
 TLS library even if it is possible to implement it at the application layer.
 TLS implementations supporting the use of exported authenticators MUST provide
 application programming interfaces by which clients and servers may request
 and verify exported authenticator messages.
 
-Given an established connection, the application SHOULD be able to call an
-"authenticate" API which takes as input:
+Given an established connection, the application SHOULD be able to call the
+following APIs:
 
- * certificate_request_context (from 0 to 255 bytes)
- * valid certificate chain for the connection and associated extensions
-(OCSP, SCT, etc.)
- * signer (either the private key associated with the certificate, or interface
-to perform private key operation)
- * signature scheme
+"request", which takes as input:
+
+* certificate_request_context (from 0 to 255 bytes)
+* set of extensions to include
+
+It returns an authenticator request.
+
+"get context", which takes as input
 
-The API returns the exported authenticator as output.
+* authenticator request
+
+It returns the certificate_request_context.
+
+"authenticate", which takes as input:
+
+ * a set of certificate chains and associated extensions
+(OCSP, SCT, etc.)
+ * a signer (either the private key associated with the certificate, or interface
+to perform private key operation) for each chain
+ * an optional authenticator request
 
-Given an established connection and an exported authenticator message, the
-application SHOULD be able to call a "validate" API that takes an exported
-authenticator as an input. If the Finished and CertificateVerify messages
-verify correctly, the API returns the following as output:
+It returns the exported authenticator as output.  The logic for selecting the
+certificates and extensions to include in the exporter SHOULD be implemented
+in the TLS library.
```

Comment:
I don't think that we need a "SHOULD" here.  Indeed, the suggested API here isn't the only potential option.  For instance, you could imagine a variant on this that has a method for asking about the properties of an authenticator request in ways that would allow for certificate selection.  You might also note that the API absent an authenticator request would not accept more than one certificate.

I'd prefer to discuss the implications of different designs here in a little more detail.  The key observation here is that the choice of certificate depends on the contents of the request.  How you manage that can vary.

Reply:
Fair commentary, I can rework this.